drivers: pwm: nrf: Update handling of `pwm` parameter and definitions of `pwm-cells`, add support for `PWM_POLARITY_INVERTED` flag

anangl opened the pull request:
This PR is a preparation for #44523.

The two nRF PWM drivers (pwm_nrfx and pwm_nrf5_sw) are aligned with other in-tree PWM drivers in the way that the `pwm` parameter of the `pwm_pin_set_cycles` function is handled. The parameter is now interpreted as a PWM channel, not an SoC pin.

Definitions of `pwm_cells` in bindings related to both the drivers are extended with `period` and `flags` cells, and support for the `PWM_POLARITY_INVERTED` flag is added in the drivers.

All affected board definitions and samples are updated accordingly.

For further details, please see the particular commit messages.

anangl committed:
drivers: pwm_nrf5_sw: Treat `pwm` parameter as PWM channel, not SoC pin

Align with other PWM drivers and treat the `pwm` parameter (described
ambiguously as "PWM pin") of the `pwm_pin_set_cycles` function as a PWM
channel, not an SoC pin. This will also make the driver consistent with
the `pwm-cells` property definition in the "nordic,nrf-sw-pwm" binding
and with related `DT_PWMS_*` macros.
The change described above requires also providing a way to specify
SoC pins that are to be assigned to the PWM channels. Hence, the commit
introduces in the "nordic,nrf-sw-pwm" binding the `channel-gpios`
property that replaces the `channel-count` one.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrfx: Treat `pwm` parameter as PWM channel, not SoC pin

Align with other PWM drivers and treat the `pwm` parameter (described
ambiguously as "PWM pin") of the `pwm_pin_set_cycles` function as a PWM
channel, not an SoC pin. This will also make the driver consistent with
the `pwm-cells` property definition in the "nordic,nrf-pwm" binding
and with related `DT_PWMS_*` macros.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: bbc_microbit: Update dts and add edge connector node

Align the board dts with the recent changes in the "nordic,nrf-sw-pwm"
binding (remove the no longer existing `channel-count` property) and
add a node representing the edge connector for convenient referring
to SoC pins connected to it.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf: Align with recent changes made to pwm_nrf5_sw driver

... and "nordic,nrf-sw-pwm" binding:

- add `channel-gpios` property with GPIO assignments for PWM channels
  to `sw_pwm` nodes
- use channel indexes instead of pin numbers in `pwms` properties that
  define PWM LEDs
- add the period and flags cells to `pwms` properties in all PWM LED
  definitions; use the commonly used period of 20 ms (giving 50 Hz)
  as a default setting

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf: Align with recent changes made to pwm_nrfx driver

... and "nordic,nrf-pwm" binding:

- use channel indexes instead of pin numbers in `pwms` properties that
  define PWM LEDs
- add the period and flags cells to `pwms` properties in all PWM LED
  definitions; use the commonly used period of 20 ms (giving 50 Hz)
  as a default setting

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrf5_sw: Add support for PWM_POLARITY_INVERTED flag

Add support for inverting of PWM channel outputs in the pwm_nrf5_sw
driver by properly handling the `PWM_POLARITY_INVERTED` flag.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: bindings: pwm: nordic: Extend pwm-cells with period and flags

... to align with what is used in most other PWM bindings.

Update PWM nodes in SoC .dtsi files accordingly.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: bbc_microbit: Align with recent changes in pwm_nrf5_sw driver

- add `channel-gpios` property with GPIO assignments for PWM channels
  to `sw_pwm` nodes
- update PWM related macros to refer to channels instead of pins
- remove no longer needed inclusion of boards/arm/bbc_microbit/board.h

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
samples: servo_motor: Align with recent changes in pwm_nrf5_sw driver

- add `channel-gpios` property with GPIO assignment for the used PWM
  channel to the `sw_pwm` node
- replace ambiguous "pin 21" in the sample's README with "pin P19"
  that uses notation from the official micro:bit documentation and
  is consistent with this pin number within the edge_connector node

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: pwm_nrfx: Add support for PWM_POLARITY_INVERTED flag

Add support for inverting of PWM channel outputs in the pwm_nrfx driver
by properly handling the `PWM_POLARITY_INVERTED` flag.
The dts properties that were used so far for inverting of the outputs
("nordic,invert" and "chX-inverted") are kept as they are needed for
setting of the initial polarity, i.e. for setting the inactive state
of the outputs before any PWM signal generation is requested for them.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

lairdjm commented on boards/arm/bt610/bt610.dts:
```diff
@@ -39,12 +39,12 @@
 		led1pwm: led_1_pwm {
-			pwms = <&pwm0 39>;
+			pwms = <&pwm0 0 PWM_MSEC(20) PWM_POLARITY_INVERTED>;
 			label = "Red PWM LED 1";
 		};
 
 		led2pwm: led_2_pwm {
-			pwms = <&pwm1 35>;
+			pwms = <&pwm1 0 PWM_MSEC(20) PWM_POLARITY_INVERTED>;
```
Should be high to be on, seems the polarity here is wrong?

anangl replied:
Right, my mistake. Thanks for catching this.

anangl replied:
Fixed.

anangl committed:
samples: bluetooth: mesh_demo: Align with changes in pwm_nrf5_sw driver

- add an overlay file for the bbc_microbit board with specification
  of GPIO that should be used as the PWM output to buzzer
- update a PWM related macro in the microbit specific code to refer
  to PWM channel instead of pin

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull reviewed (approved):
LGTM, thanks for these changes! Note I haven’t reviewed in detail boards conversion/updates.

henrikbrixandersen reviewed (approved):
Thanks!
